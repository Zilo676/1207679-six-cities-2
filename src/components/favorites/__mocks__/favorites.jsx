import React from "react";

const Favorites = () => {
  return <></>;
};

export {Favorites};
export default Favorites;
