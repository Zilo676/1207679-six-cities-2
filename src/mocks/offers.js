const offers = [
  {
    id: `799663af-e554-49c5-b73e-d61b36b503f0`,
    type: `Apartament`,
    price: 120,
    description: `Beautiful & luxurious apartment at great location`,
    coordinates: [52.3909553943508, 4.85309666406198],
    raiting: 93,
    city: `Amsterdam`,
  },
  {
    id: `4a6cf5ff-150e-40f2-baf8-820461958a61`,
    type: `Private room`,
    price: 80,
    description: `Wood and stone place`,
    coordinates: [52.369553943508, 4.85309666406198],
    raiting: 80,
    city: `Amsterdam`,
  },
  {
    id: `0dcee833-d3f2-4f12-9559-a2775ca7f733`,
    type: `Apartament`,
    price: 132,
    description: `Canal View Prinsengracht`,
    coordinates: [52.3909553943508, 4.929309666406198],
    raiting: 80,
    city: `Amsterdam`,
  },
  {
    id: `f5035346-5f36-480f-9401-b8f5dc8186c6`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [52.3809553943508, 4.939309666406198],
    raiting: 100,
    city: `Amsterdam`,
  },
  {
    id: `76119596-1f8c-4df9-bb83-ce109ef8f3af`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [48.864716, 2.349014],
    raiting: 100,
    city: `Paris`,
  },
  {
    id: `76119596-1f8c-4df9-bb83-ce109ef8f3af`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [50.935173, 6.953101],
    raiting: 100,
    city: `Colonge`,
  },
  {
    id: `76119596-1f8c-4df9-bb83-ce109ef8f3af`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [50.85045, 4.34878],
    raiting: 100,
    city: `Brussels`,
  },
  {
    id: `76119596-1f8c-4df9-bb83-ce109ef8f3af`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [53.551086, 9.993682],
    raiting: 100,
    city: `Hamburg`,
  },
  {
    id: `76119596-1f8c-4df9-bb83-ce109ef8f3af`,
    type: `Apartament`,
    price: 180,
    description: `Nice, cozy, warm big bed apartment`,
    coordinates: [51.2277411, 6.7734556],
    raiting: 100,
    city: `Dusseldorf`,
  }
];

export {offers};
