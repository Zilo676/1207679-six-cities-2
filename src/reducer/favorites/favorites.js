import {
  Operation as hotelsOperation
} from '../hotels/hotels';

const initialState = {
  favorites: [],
};

const ActionType = {
  LOAD_FAVORITES: `LOAD_FAVORITES`,
};

const ActionCreator = {
  loadFavorites: (favorites) => {
    return {
      type: ActionType.LOAD_FAVORITES,
      payload: favorites,
    };
  },
};

const Operation = {
  loadFavorites: () => (dispatch, _getState, api) => {
    return api.get(`/favorite`)
      .then((response) => {
        dispatch(ActionCreator.loadFavorites(response.data));
      });
  },

  changeFavoriteStatus: (hotelId, status) => (dispatch, _getState, api) => {
    return api.post(`/favorite/${hotelId}/${status}`)
      .then((response) => {
        dispatch(ActionCreator.loadFavorites(response.data));
        dispatch(hotelsOperation.loadHotels());
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_FAVORITES:
      return Object.assign({}, state, {
        favorites: action.payload,
      });
  }

  return state;
};

export {
  ActionCreator,
  ActionType,
  Operation,
  reducer,
};
