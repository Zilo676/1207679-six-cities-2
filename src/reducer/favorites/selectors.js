import NameSpace from '../name-spaces';
import {createSelector} from 'reselect';

import {getAllCities} from '../city/selectors';

const NAME_SPACE = NameSpace.FAVORITES;

const getFavorites = (state) => state[NAME_SPACE].favorites;

const getFavoritesByCities = createSelector(
    getFavorites,
    getAllCities,
    (favorites, cities) => {
      let result = {};
      cities.forEach((city) => {
        result[city] = [];
        if (favorites) {
          favorites.forEach((it) => {
            if (it.city.name === city) {
              result[city].push(it);
            }
          });
        }
      });
      return result;
    }
);

export {
  getFavorites,
  getFavoritesByCities
};
