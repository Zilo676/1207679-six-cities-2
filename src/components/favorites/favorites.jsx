import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

import {Path} from '../../api';

import {OfferList} from '../offer-list/offer-list.jsx';
import Header from '../header/header.jsx';
import {FavoritesEmpty} from '../favorites-empty/favorites-empty.jsx';

import {getFavoritesByCities, getFavorites} from '../../reducer/favorites/selectors.js';
import {getAllCities} from '../../reducer/city/selectors';
import {ActionCreator} from '../../reducer/city/city';

const Favorites = (props) => {
  const {allCities, favorites, allFavorites, onClick} = props;
  return (
    <React.Fragment>
      <div className="page__favorites-container container">
        <Header />
        <h1 className="favorites__title">Saved listing</h1>
        <ul className="favorites__list">

          {allFavorites.length === 0 && <FavoritesEmpty />}
          {(allFavorites.length !== 0 &&
            allCities.map((city, i) => {
              return (favorites[city].length > 0 &&
                <li key={city + i} className="favorites__locations-items">
                  <div className="favorites__locations locations locations--current">
                    <div className="locations__item">
                      <Link className="locations__item-link" to={Path.MAIN_PAGE}><span onClick={() => onClick(city)}>{city}</span></Link>
                    </div>
                  </div>
                  <div className="favorites__places">
                    <OfferList offers={favorites[city]} onActiveItem={() => {}} onNotActiveItem={() => {}} />
                  </div>
                </li>
              );
            })
          )}
        </ul>
      </div>
      <footer className="footer container">
        <Link to={Path.MAIN_PAGE} className="footer__logo-link">
          <img className="footer__logo" src="/img/logo.svg" alt="6 cities logo" width="64" height="33" />
        </Link>
      </footer>
    </React.Fragment>
  );
};

Favorites.propTypes = {
  allCities: PropTypes.array,
  favorites: PropTypes.shape(),
  onClick: PropTypes.func,
  allFavorites: PropTypes.shape(),
};

const mapStateTpProps = (state, ownProps) => Object.assign({}, ownProps, {
  allCities: getAllCities(state),
  favorites: getFavoritesByCities(state),
  allFavorites: getFavorites(state),
});

const mapDispatchToProps = (dispatch) => ({
  onClick: (city) => {
    dispatch(ActionCreator.setCity(city));
  },
});

export default connect(mapStateTpProps, mapDispatchToProps)(Favorites);

export {Favorites};
