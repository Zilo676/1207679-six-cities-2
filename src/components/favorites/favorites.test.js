import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';

import {Favorites} from './favorites.jsx';
import offers from '../../mock/offers';

jest.mock(`../header/header.jsx`);
jest.mock(`../offer-list/offer-list.jsx`);

it(`Correctly rendered`, () => {
  const mockFunc = jest.fn();
  const tree = renderer.create(<BrowserRouter><Favorites
    allCities={[`Amsterdam`]}
    favorites={{Amsterdam: [offers[0]]}}
    onClick={mockFunc}
    allFavorites={offers}
  />
  </BrowserRouter>).toJSON();
  expect(tree).toMatchSnapshot();
});
