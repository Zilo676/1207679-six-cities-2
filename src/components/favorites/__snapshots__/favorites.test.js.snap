// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly rendered 1`] = `
Array [
  <div
    className="page__favorites-container container"
  >
    <h1
      className="favorites__title"
    >
      Saved listing
    </h1>
    <ul
      className="favorites__list"
    >
      <li
        className="favorites__locations-items"
      >
        <div
          className="favorites__locations locations locations--current"
        >
          <div
            className="locations__item"
          >
            <a
              className="locations__item-link"
              href="/"
              onClick={[Function]}
            >
              <span
                onClick={[Function]}
              >
                Amsterdam
              </span>
            </a>
          </div>
        </div>
        <div
          className="favorites__places"
        />
      </li>
    </ul>
  </div>,
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>,
]
`;
